--- res/res_rtp_asterisk.c.orig	2014-12-08 15:53:43.032358579 +0100
+++ res/res_rtp_asterisk.c	2014-12-08 15:54:15.524358525 +0100
@@ -47,6 +47,7 @@
 #include <openssl/bio.h>
 #endif
 
+#undef USE_PJPROJECT
 #ifdef USE_PJPROJECT
 /* Asterisk discourages the use of bzero in favor of memset, in fact if you try to use bzero it will tell you to use memset. As a result bzero has to be undefined
  * here since it is used internally by pjlib. The only other option would be to modify pjlib... which won't happen. */
